File: King of Modules/Backups/16-03-26 v2.0.1 --- Python MT5 e Backtest/Strategies/KamaTrend.py
import pandas as pd
 
import MetaTrader5 as mt5
 
from Config import cfg, Signal
from Indicators import (
    compute_kama,
    compute_atr,
    get_atr_value,
)
 
KAMA_FLIP_LOOKBACK = 3   # mirrors #define KAMA_FLIP_LOOKBACK 3
 
 
# ---------------------------------------------------------------------------
# KAMA Flip Detection — GetKamaFlip
# ---------------------------------------------------------------------------
 
def _get_kama_flip(rates: pd.DataFrame):
    """
    Scans the last KAMA_FLIP_LOOKBACK+1 closed regime values for a flip.
    Returns (bullish_flip, bearish_flip) booleans, or None on failure.
 
    Mirrors GetKamaFlip():
        CopyBuffer(g_handleKAMA, 1, offset=1, count=LOOKBACK+1)
        loop i=0..LOOKBACK-1:
          current  = regimeBuf[i]       (newer)
          previous = regimeBuf[i + 1]   (older)
    """
    _, regime = compute_kama(rates["close"])
 
    needed = KAMA_FLIP_LOOKBACK + 2   # +1 for lookback window, +1 to skip live bar
    if len(regime) < needed:
        return None
 
    # offset=1: skip the live (forming) bar
    # We take the last (KAMA_FLIP_LOOKBACK + 1) closed regime values
    window = regime.iloc[-(KAMA_FLIP_LOOKBACK + 2):-1].values
 
    # In MQL5 CopyBuffer with array filled oldest→newest:
    # index 0 = oldest, index LOOKBACK = newest of the window
    # Loop: i=0 → current=buf[0], previous=buf[1]  (MQL5 newest-first convention)
    # Python: we reverse so index 0 = newest, index -1 = oldest
    # Then: current = window_rev[i], previous = window_rev[i+1]
 
    window_rev = window[::-1]   # newest first
 
    bullish_flip = False
    bearish_flip = False
 
    for i in range(KAMA_FLIP_LOOKBACK):
        current  = int(round(window_rev[i]))
        previous = int(round(window_rev[i + 1]))
 
        if previous == 1 and current == 2:
            bullish_flip = True
        if previous == 2 and current == 1:
            bearish_flip = True
 
    return bullish_flip, bearish_flip
 
 
# ---------------------------------------------------------------------------
# ATR Expansion Filter — PassATRExpansionFilter
# ---------------------------------------------------------------------------
 
def _pass_atr_expansion_filter(rates: pd.DataFrame) -> bool:
    """
    Allow entry only when:
      ATR(14) is rising  (now > prev bar)
      AND ATR(14) > ATR(50) * multiplier
 
    Mirrors PassATRExpansionFilter().
    """
    if not cfg.use_atr_kt_trend_filter:
        return True
 
    atr_fast_series = compute_atr(rates, cfg.atr_period)
    if len(atr_fast_series) < 4:
        return False
 
    # offset=1: bar[1] and bar[2] (both closed)
    atr_fast_now  = atr_fast_series.iloc[-2]   # bar[1]
    atr_fast_prev = atr_fast_series.iloc[-3]   # bar[2]
 
    if atr_fast_now <= atr_fast_prev:
        return False
 
    atr_slow = get_atr_value(rates, 50)
    if atr_slow is None:
        return False
 
    return atr_fast_now > atr_slow * cfg.atr_kt_trend_multiplier
 
 
# ---------------------------------------------------------------------------
# ADX Trend Filter — PassADXTrendFilter
# ---------------------------------------------------------------------------
 
def _compute_adx(rates: pd.DataFrame, period: int) -> pd.Series:
    """
    Wilder-smoothed ADX matching MQL5 iADX.
    Returns the ADX line (buffer 0 in MQL5).
    """
    high  = rates["high"]
    low   = rates["low"]
    close = rates["close"]
 
    up_move   = high.diff()
    down_move = -low.diff()
 
    plus_dm  = up_move.where((up_move > down_move) & (up_move > 0), 0.0)
    minus_dm = down_move.where((down_move > up_move) & (down_move > 0), 0.0)
 
    tr = pd.concat(
        [(high - low),
         (high - close.shift()).abs(),
         (low  - close.shift()).abs()],
        axis=1
    ).max(axis=1)
 
    atr      = tr.ewm(alpha=1 / period, adjust=False).mean()
    plus_di  = 100 * plus_dm.ewm(alpha=1 / period, adjust=False).mean() / atr
    minus_di = 100 * minus_dm.ewm(alpha=1 / period, adjust=False).mean() / atr
 
    dx  = (100 * (plus_di - minus_di).abs() /
           (plus_di + minus_di).replace(0, float("nan")))
    adx = dx.ewm(alpha=1 / period, adjust=False).mean()
    return adx
 
 
def _pass_adx_trend_filter(rates: pd.DataFrame) -> bool:
    """
    Allow entry when ADX is rising AND ADX < cfg.adx_comparison.
    Note: the original comment shows experimentation (< 24 / > 19.5 / > 18).
    The active code is: adxNow < InpADXComparision.
 
    Mirrors PassADXTrendFilter().
    """
    if not cfg.use_adx_filter:
        return True
 
    adx = _compute_adx(rates, cfg.adx_period)
    if len(adx) < cfg.adx_period + 3:
        return False
 
    adx_now  = adx.iloc[-2]   # bar[1]
    adx_prev = adx.iloc[-3]   # bar[2]
 
    if adx_now <= adx_prev:
        return False
 
    return adx_now < cfg.adx_comparison
 
 
# ---------------------------------------------------------------------------
# Public signal function — SignalKamaTrend (orchestrator)
# ---------------------------------------------------------------------------
 
def signal_kama_trend(rates: pd.DataFrame) -> Signal:
    """
    Main entry point. Call once per new closed bar.
 
    Returns Signal.BUY / Signal.SELL / Signal.NONE.
    """
    flip_result = _get_kama_flip(rates)
    if flip_result is None:
        return Signal.NONE
 
    bullish_flip, bearish_flip = flip_result
 
    if not bearish_flip and not bullish_flip:
        return Signal.NONE
 
    if not _pass_atr_expansion_filter(rates):
        return Signal.NONE
 
    if not _pass_adx_trend_filter(rates):
        return Signal.NONE
 
    # Bearish flip takes priority (matches MQL5 evaluation order)
    if bearish_flip:
        return Signal.SELL
 
    if bullish_flip:
        return Signal.BUY
 
    return Signal.NONE

File: King of Modules/Zorro/Backups/19-03-2026 v1.0.1/Strategies/KamaTrend.c
#ifndef KAMA_TREND_C
#define KAMA_TREND_C

#include "../Indicators.c"

#define KAMA_FLIP_LOOKBACK 3


// ============================================================
//  GetKamaFlip
//  Scans the last KAMA_FLIP_LOOKBACK bar-pairs for a regime flip.
//  bullishFlip → any bar in the window transitioned bearish→bullish
//  bearishFlip → any bar in the window transitioned bullish→bearish
//
//  kamaColorSeries() index:
//    [1] = most recent closed bar (bar 1)
//    [2] = bar 2, etc.
//  Pairs examined: ([1],[2]), ([2],[3]), ([3],[4])
//    i.e. current=cs[i], previous=cs[i+1] for i=1..KAMA_FLIP_LOOKBACK
// ============================================================
static int GetKamaFlip(int* bullishFlip, int* bearishFlip)
{
    var* cs = kamaColorSeries();

    *bullishFlip = 0;
    *bearishFlip = 0;

    for(int i = 1; i <= KAMA_FLIP_LOOKBACK; i++)
    {
        int current  = (int)cs[i];
        int previous = (int)cs[i + 1];

        // bearish→bullish: 1→0
        if(previous == 1 && current == 0) *bullishFlip = 1;
        // bullish→bearish: 0→1
        if(previous == 0 && current == 1) *bearishFlip = 1;
    }

    return 1;
}


// ============================================================
//  PassATRExpansionFilter
//  Allows entry only when:
//    (a) ATR(14) is rising vs. previous bar  AND
//    (b) ATR(14) > ATR(50) * InpATRKTTrendMultiplier
//  (expanding volatility / trending regime).
// ============================================================
static int PassATRExpansionFilter()
{
    if(!InpUseATRKTTrendFilter)
        return 1;

    var* atrFastSeries = series(ATR(InpATRPeriod));
    var  atrFastNow    = atrFastSeries[0];
    var  atrFastPrev   = atrFastSeries[1];

    // ATR must be rising
    if(atrFastNow <= atrFastPrev)
        return 0;

    var atrSlow = series(ATR(50))[0];

    return (atrFastNow > atrSlow * InpATRKTTrendMultiplier);
}


// ============================================================
//  PassADXTrendFilter
//  Allows entry only when:
//    (a) ADX is rising vs. previous bar  AND
//    (b) ADX < InpADXComparision (not yet in overextended trend)
// ============================================================
static int PassADXTrendFilter()
{
    if(!InpUseADXFilter)
        return 1;

    var* adxSeries = series(ADX(InpADXPeriod));
    var  adxNow    = adxSeries[0];
    var  adxPrev   = adxSeries[1];

    if(adxNow <= adxPrev)
        return 0;

    return (adxNow < InpADXComparision);
}


// ============================================================
//  SignalKamaTrend  (public)
//  Fires SIGNAL_BUY on bullish KAMA flip, SIGNAL_SELL on bearish,
//  subject to ATR expansion and ADX trend strength filters.
// ============================================================
int SignalKamaTrend()
{
    int bullishFlip = 0, bearishFlip = 0;
    if(!GetKamaFlip(&bullishFlip, &bearishFlip))
        return SIGNAL_NONE;

    if(!bullishFlip && !bearishFlip)
        return SIGNAL_NONE;

    if(!PassATRExpansionFilter())
        return SIGNAL_NONE;

    if(!PassADXTrendFilter())
        return SIGNAL_NONE;

    if(bearishFlip) return SIGNAL_SELL;
    if(bullishFlip) return SIGNAL_BUY;

    return SIGNAL_NONE;
}

#endif // KAMA_TREND_C
